### PR TITLE
[Compute] `az vmss create`: Add `--enable-auto-os-upgrade` parameter to support AutomaticOSUpgrade while creating

### LINKED ISSUE #28493: Enable Auto Image Upgrade while Creating new VMSS - CLI Change Request
### Preconditions

- [X] No need to upgrade Python SDK or the Python SDK is ready.

### Related command

az vmss create --name myScaleSet --resource-group myResourceGroup --set UpgradePolicy.AutomaticOSUpgradePolicy.EnableAutomaticOSUpgrade=true

### Resource Provider

Compute Resource Provider

### Description of Feature or Work Requested

Allow customers to enable Auto OS Image Upgrades on their VM Scale Sets for 1P Gallery using CLI. This capability is already available on Powershell and API. The 1P Gallery is created as part of Project Standard to ensure 1P assets have an improved security posture. In order for customers to migrate to the gallery they need to enable Auto OS Image Upgrades (https://learn.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-automatic-upgrade?toc=%2fazure%2fvirtual-machines%2ftoc.json).

### Minimum API Version Required

Use Compute API version 2018-10-01 or higher

### Swagger PR link / SDK link

https://learn.microsoft.com/en-us/cli/azure/vmss?view=azure-cli-latest#az-vmss-create

### Request Example

_No response_

### Target Date

2024-04-30

### PM Contact

Maulik Shah

### Engineer Contact

Deepali Gupta

### Additional context

This feature is required for the 1P Image Gallery as well as for other customers that use the Auto OS Image Upgrade capability to remain secured.

### PR DESCRIPTION
**Pull request by ReaNAiveD**
**Related command**

* `az vmss create --enable-auto-os-upgrade`

**Description**

Resolve #28493

Allow customers to enable Auto OS Image Upgrades on their VM Scale Sets for 1P Gallery using CLI. This capability is already available on Powershell and API. The 1P Gallery is created as part of Project Standard to ensure 1P assets have an improved security posture. In order for customers to migrate to the gallery they need to enable Auto OS Image Upgrades (https://learn.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-automatic-upgrade?toc=%2fazure%2fvirtual-machines%2ftoc.json).

**Testing Guide**

```bash
# Create VMSS with AutomaticOSUpgrade
vmss create --debug -n MyVMSS -g MyResourceGroup --image MicrosoftWindowsServer:WindowsServer:2022-Datacenter:latest --upgrade-policy-mode Automatic --enable-auto-os-upgrade --orchestration-mode Uniform --load-balancer myLoadBalancer --admin-username TestUser --admin-password Test123456789; --vm-sku Standard_A1 --health-probe MyProbe
```

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Compute

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 1]
```diff
@@ -134,6 +134,8 @@ def load_arguments(self, _):
     # The `Standard` is used for backward compatibility to allow customers to keep their current behavior after changing the default values to Trusted Launch VMs in the future.
     t_security = [x.value for x in self.get_models('SecurityTypes') or []] + [COMPATIBLE_SECURITY_TYPE_VALUE]
     security_type = CLIArgumentType(arg_type=get_enum_type(t_security), min_api='2020-12-01', help='Specify the security type of the virtual machine.')
+    enable_auto_os_upgrade_type = CLIArgumentType(arg_type=get_three_state_flag(), min_api='2018-10-01',
+                                                  help='Indicates whether OS upgrades should automatically be applied to scale set instances in a rolling fashion when a newer version of the OS image becomes available.')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:138]: Please use the first-person voice in the help message
```suggestion
                                                  help='Indicate whether OS upgrades should automatically be applied to scale set instances in a rolling fashion when a newer version of the OS image becomes available.')
```

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 2]
```diff
@@ -1307,6 +1307,16 @@ def build_vmss_resource(cmd, name, computer_name_prefix, location, tags, overpro
...
+    if upgrade_policy_mode and cmd.supported_api_version(min_api='2018-10-01',
+                                                         operation_group='virtual_machine_scale_sets'):
+        vmss_properties['upgradePolicy']['automaticOSUpgradePolicy'] = {}
+        automatic_os_upgrade_policy = vmss_properties['upgradePolicy']['automaticOSUpgradePolicy']
+
+        if enable_auto_os_upgrade is not None:
+            automatic_os_upgrade_policy['enableAutomaticOSUpgrade'] = enable_auto_os_upgrade
+
+        if not automatic_os_upgrade_policy:
+            del automatic_os_upgrade_policy
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:1319]: Why is this logic needed?

Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:1319]: Why is this logic needed?

Comment (ReaNAiveD) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:1319]: I emulated the logic of the `rolling_upgrade_policy` part, which is a concept similar to `automatic_os_upgrade_policy`.

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:1319]: OK, I guess its purpose is to not pass it to the server when `rolling_upgrade_policy` is not available

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**